### PR TITLE
Allow works to have a small number of legacy licenses

### LINKED ISSUE #1047: Support a list of licenses that can’t be chosen but can be displayed
This is necessary for migrating content from Hydrus that has licenses that we are not supporting in H2.

When items with unsupported licenses are migrated, the depositor should see the license they previously chosen, e.g. a CC 3.0 license, displayed on the item show page. When the user versions this item, they will be required to choose one of the valid H2 license options in order to be able to deposit the new version.

Content in draft that has a license assigned will be required to choose a new license, unless that license was set by the collection manager already.

### PR DESCRIPTION
**Pull request by mjgiarlo**
Fixes #1047

## Why was this change made?

This change is intended to help migrate Hydrus 1 data into Hydrus 2, and to display migrated works with their current (CC 3.0) licenses. It allows a small set of displayable licenses to show both in the work show page and on the work form. Before a new version of such a work may be deposited again, a new H2 license must be selected.

Includes:

* Remove license validation from work model
  * We are already validating licenses in the work form so this is not necessary and is counter to our approach to validation where we favor form validation over modal validation. This also allows us to migrate *any* license from Hydrus 1.0.

### Screenshot: Show legacy license in show page

license_show

### Screenshot: Display legacy license in form

selected_license

### Screencast: Show dropdown behavior in form

Peek 2021-02-12 14-58

## How was this change tested?

CI, local browser

## Which documentation and/or configurations were updated?

None

### CODE REVIEW

File: app/components/works/detail_component.rb
Comment (mjgiarlo) [inline app/components/works/detail_component.rb:65]: This is an internal refactoring to use the public interface of the License model, preferring message-sending to using constants. Has no effect outside the component.

File: app/components/works/license_component.html.erb
Comment (mjgiarlo) [inline app/components/works/license_component.html.erb:8]: Pass the license to the method so it knows whether to show the merely displayable licenses as well as the selectable ones. (Don't show them unless there's a legacy license.)

File: app/models/license.rb
Comment (mjgiarlo) [inline app/models/license.rb:10]: Rename to clarify distinction between selectable licenses and displayable ones.

File: app/models/work.rb
Comment (mjgiarlo) [inline app/models/work.rb:23]: We already have these validations in the form, which is more consistent with our validation philosophy. No need to have them here.

**Comment by peetucket** [comment]
LGTM, not sure if you want another approver before merging.

**Review by peetucket — APPROVED** [review]

File: app/models/license.rb
Comment (jcoyne) [inline app/models/license.rb:61]: I don't think we want to add this here, this would be in the selectable list. We want the user to select a new valid license if they make a change.

File: app/models/work.rb
Comment (jcoyne) [reply app/models/work.rb:23]: I think it's a good idea to have them in both places. We want to ensure the licences in the model is SELECTABLE_LICENSES or DISPLAYABLE_LICESES, but we want those in the form to only be the SELECTABLE_LICENSES.  We want to make sure we aren't importing any invalid data (which bypasses the form).

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/models/license.rb
Comment (mjgiarlo) [reply app/models/license.rb:61]: @jcoyne It shouldn't because of the check on line 58. This only appends displayable licenses if one of them is already selected.

Comment (jcoyne) [reply app/models/license.rb:61]: 👍

File: app/models/work.rb
Comment (mjgiarlo) [reply app/models/work.rb:23]: @jcoyne I'm open to making this change and it's easy enough. This leaves me fuzzy on the following question: how do we decide where a validation ought to live? What's the guideline?

I hear you regarding not wanting bogus licenses in the database. If that's true, doesn't that imply we want a bunch more of our validations down inthe models? (Less important sidebar: And if they move to the models, would we even need them in the forms anymore?) It'd be great to get a shared sense of this, so I'm copying @sul-dlss/infrastructure-team

Comment (peetucket) [reply app/models/work.rb:23]: I am interested in this question too, having previously moved validations around between forms and models.  Our use of draft vs regular forms to deal with validations needed sometimes but not other times just adds to the complexity (and maybe the original driver behind moving most if not all validation outside the model).

Comment (jcoyne) [reply app/models/work.rb:23]: I think you want to validate things you put in the database (for migrations, etc).  You also want to validate some of the same things on the form (and possibly some more things).  I don't think there is a simple rule of thumb for this. Validations are complex and context dependent.

Comment (mjgiarlo) [reply app/models/work.rb:23]: Yeah, this is rather muddy for me, still. I'm happy to embrace the muddiness and rejigger this, though.

Comment (mjgiarlo) [reply app/models/work.rb:23]: Done.

**Comment by mjgiarlo** [comment]
@jcoyne this is ready for another look.

File: .rubocop.yml
Comment (mjgiarlo) [inline .rubocop.yml:253]: Rubocop changed while I was working on this PR, so I included the new rules here.

File: app/models/work.rb
Comment (jcoyne) [inline app/models/work.rb:22]: I'm not sure why these disappeared in this PR . can they be restored?   Maybe remove in a different PR?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/models/work.rb
Comment (mjgiarlo) [reply app/models/work.rb:22]: Ah, these disappeared because of the conversation we had re: "validation in forms or models" we had on take 1 of the PR. So, I will restore these and push.

Comment (mjgiarlo) [reply app/models/work.rb:22]: @jcoyne Fixed. One more look?

**Review by jcoyne — APPROVED** [review]

**Pull request merged**
